gio/src/task.rs

```diff
@@ -60,10 +85,188 @@ impl Task {
+            f(&from_glib_borrow(this))
+        }
+        unsafe {
+            let f: Box_<F> = Box_::new(f);
+            connect_raw(
+                self.as_ptr() as *mut _,
+                b"notify::completed\0".as_ptr() as *const _,
+                Some(transmute::<_, unsafe extern "C" fn()>(
+                    notify_completed_trampoline::<V, F> as *const (),
+                )),
+                Box_::into_raw(f),
+            )
+        }
+    }
+
+    #[doc(alias = "g_task_return_value")]
+    #[doc(alias = "g_task_return_boolean")]
+    #[doc(alias = "g_task_return_int")]
+    #[doc(alias = "g_task_return_pointer")]
+    #[doc(alias = "g_task_return_error")]
+    pub fn return_result(&self, result: Result<&V, glib::Error>) {
+        #[cfg(not(feature = "v2_64"))]
+        unsafe extern "C" fn value_free(value: *mut libc::c_void) {
+            let _: glib::Value = from_glib_full(value as *mut glib::gobject_ffi::GValue);
+        }
+
+        match result {
+            #[cfg(feature = "v2_64")]
+            Ok(v) => unsafe {
+                ffi::g_task_return_value(
+                    self.to_glib_none().0,
+                    v.to_value().to_glib_full() as *mut _,
+                )
+            },
+            #[cfg(not(feature = "v2_64"))]
+            Ok(v) => unsafe {
+                ffi::g_task_return_pointer(
+                    self.to_glib_none().0,
+                    v.to_value().to_glib_full() as *mut _,
+                    Some(value_free),
+                )
+            },
+            Err(e) => unsafe {
+                ffi::g_task_return_error(self.to_glib_none().0, e.to_glib_full() as *mut _);
+            },
+        }
+    }
+
+    #[doc(alias = "g_task_propagate_value")]
+    #[doc(alias = "g_task_propagate_boolean")]
+    #[doc(alias = "g_task_propagate_int")]
+    #[doc(alias = "g_task_propagate_pointer")]
+    pub fn propagate(self) -> Result<V, glib::Error> {
+        unsafe {
+            let mut error = ptr::null_mut();
+
+            #[cfg(feature = "v2_64")]
+            {
+                let mut value = glib::Value::uninitialized();
+                ffi::g_task_propagate_value(
+                    self.to_glib_none().0,
+                    value.to_glib_none_mut().0,
+                    &mut error,
+                );
+
+                if error.is_null() {
+                    Ok(V::from_value(&value))
+                } else {
+                    Err(from_glib_full(error))
+                }
+            }
+
+            #[cfg(not(feature = "v2_64"))]
+            {
+                let value = ffi::g_task_propagate_pointer(self.to_glib_none().0, &mut error);
+
+                if error.is_null() {
+                    let value = Option::<glib::Value>::from_glib_full(
+                        value as *mut glib::gobject_ffi::GValue,
+                    )
+                    .unwrap_or_else(|| glib::Value::from_type(glib::types::Type::UNIT));
```

Comment:
What are you trying to do here? If the return value is `NULL` then this should probably just panic or return an `Err`?

Reply:
I.e. it can only be `NULL` if nothing was returned yet, and then it's a programmer error to call this function so it should probably just panic.

Reply:
It was handled in the old code, hence my doubts...

Reply:
You wrote the old code ;)